prov/rxm/src/rxm.h

```diff
@@ -116,6 +116,137 @@ extern size_t rxm_msg_tx_size;
 extern size_t rxm_msg_rx_size;
 extern size_t rxm_def_univ_size;
 
+/*
+ * Connection Map
+ */
+
+#define RXM_CMAP_IDX_BITS OFI_IDX_INDEX_BITS
+
+enum rxm_cmap_signal {
+	RXM_CMAP_FREE,
+	RXM_CMAP_EXIT,
+};
+
+enum rxm_cmap_state {
+	RXM_CMAP_IDLE,
+	RXM_CMAP_CONNREQ_SENT,
+	RXM_CMAP_CONNREQ_RECV,
+	RXM_CMAP_ACCEPT,
+	RXM_CMAP_CONNECTED_NOTIFY,
+	RXM_CMAP_CONNECTED,
+	RXM_CMAP_SHUTDOWN,
+};
+
+enum rxm_cmap_reject_flag {
+	RXM_CMAP_REJECT_GENUINE,
+	RXM_CMAP_REJECT_SIMULT_CONN,
+};
+
+struct rxm_cmap_handle {
+	struct rxm_cmap *cmap;
+	enum rxm_cmap_state state;
+	/* Unique identifier for a connection. Can be exchanged with a peer
+	 * during connection setup and can later be used in a message header
+	 * to identify the source of the message (Used for FI_SOURCE, RNDV
+	 * protocol, etc.) */
+	uint64_t key;
+	uint64_t remote_key;
+	fi_addr_t fi_addr;
+	struct rxm_cmap_peer *peer;
+};
+
+struct rxm_cmap_peer {
+	struct rxm_cmap_handle *handle;
+	struct dlist_entry entry;
+	uint8_t addr[];
+};
+
+typedef struct rxm_cmap_handle*
+(*rxm_cmap_alloc_handle_func)(struct rxm_cmap *cmap);
+typedef void (*rxm_cmap_handle_func)(struct rxm_cmap_handle *handle);
+typedef int (*rxm_cmap_connect_func)(struct util_ep *ep,
+				     struct rxm_cmap_handle *handle,
+				     const void *addr, size_t addrlen);
+typedef void *(*rxm_cmap_thread_func)(void *arg);
+typedef int (*rxm_cmap_signal_func)(struct util_ep *ep, void *context,
+				    enum rxm_cmap_signal signal);
+typedef int (*rxm_cmap_cleanup_func)(void *arg);
+
+struct rxm_cmap_attr {
+	void 				*name;
+	/* user guarantee for serializing access to cmap objects */
+	uint8_t				serial_access;
+	rxm_cmap_alloc_handle_func 	alloc;
+	rxm_cmap_handle_func 		close;
+	rxm_cmap_handle_func 		save_conn;
+	rxm_cmap_handle_func 		close_saved_conn;
+	rxm_cmap_handle_func 		free;
+	rxm_cmap_connect_func 		connect;
+	rxm_cmap_handle_func		connected_handler;
+	rxm_cmap_thread_func		cm_thread_func;
+	rxm_cmap_cleanup_func		cleanup;
+	rxm_cmap_signal_func		signal;
+	rxm_cmap_handle_func		av_updated_handler;
+};
```

Comment:
Are callback functions needed?  Do these ever get set to anything other than the same function now?

Reply:
I have one more patch in the works which removes all function pointers.